### ISSUE TITLE
Usability feedback on DurableOrchestrationClient

### ISSUE THREAD

**Issue by olitomlinson** [Azure/azure-functions-durable-extension#1220]
# Feedback 1

I would find it very useful to include an additional enum in `OverridableExistingInstanceStates` that does not allow restarting an orchestration if it is already in a `Completed` status.

**Why?**

Because once I have an orchestration that has reached a `Completed` status, I _never_ want to restart it. When starting an orchestration via `ServiceBusTrigger` with at-least once semantics, if service bus redelivers the message (which can happen for many reasons) It would also restart the orchestration providing the orchestration had completed _faster_ than the `ServiceBusTrigger` redelivers the message, which in my use-case does happen.

Essentially this could look something like :

`"OverridableExistingInstanceStates" : "FailureAndTerminatedOnly"`

Was there a design decision made to not support an array of states which would allow the user to define the states that are overridable? Something like :

`"OverridableExistingInstanceStates" : [ "Failed", "Terminated" ]`

---

# Feedback 2

There are a few places where the `DurableOrchestrationClient` throws exceptions in scenarios where I would like them to be suppressed, described below in the two examples.

### Actual
**Given** an orchestration instance `A` is `Completed`, 
**When** an external system attempts to raise an event for orchestration `A`
**Then** an exception is thrown. 

### Exepected
**Given** an orchestration instance `A` is `Completed`, 
**When** an external system attempts to raise an event for orchestration `A`
**Then** an exception is **not** thrown

### Workaround 
At present I use a clunky extension method on `DurableOrchestrationClient` that 
suppresses this **particular** exception. It is ugly because it needs to parse the exception message which could change between versions of DF. (strings designed for humans don't make good APIs to rely on)

```
  public static async Task SupressTerminalStateExceptionWhenRaisingEvent(this DurableOrchestrationClientBase client, Func<DurableOrchestrationClientBase, Task> action)
  {
      try
      {
          await action(client);
      }
      catch (InvalidOperationException ex) when (ex.Message.ToLower().Contains("cannot raise event"))
      {
          // Supress this exception as the orchestration has likely already 
          // received this event and is now terminal (completed, failed etc)
      }
  }
```

```
      await client.SupressTerminalStateExceptionWhenRaisingEvent(_ 
          => _.RaiseEventAsync(instanceId, eventName, payload));
```

### solution

Make exceptions first-class so that they can be caught and filtered without relying on parsing strings designed for human readability. 

### alternate solution

I understand that this expected behavior might not be desirable for all use-cases and breaking-changes need to be considered.

Maybe include some `options` that can control the exception behavior on the call to `RaiseEventAsync` such as :

```
   var options = new RaiseEventOptions()
   {
      suppressExceptionWhenInState = { OrchestrationRuntimeStatus.Completed }
   };
   await client.RaiseEventAsync(... , ..., ...., options);  
```

## Example 2
 
When starting an orchestration via `ServiceBusTrigger` with at-least once semantics, its possible that your orchestration gets started again accidentally. This throws an exception if the orchestration is already running. I followed the same suppression pattern to ensure only this specific exception gets caught and swallowed.

```
   public static async Task SupressExceptionWhenOrchestrationIsAlreadyRunning(this 
   DurableOrchestrationClientBase client, Func<DurableOrchestrationClientBase, Task> action)
        {
            try
            {
                await action(client);
            }
            catch (InvalidOperationException ex) when (ex.Message.ToLower().Contains("an orchestration instance with the status"))
            {
                // https://github.com/Azure/durabletask/blob/effb469a04480d35935c67bf07c50e902cb3e9be/src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs#L1296
                // Supress this exception as the orchestration is already running.
                // We don't need to start it again at this point.
            }
        }
```

```
    await client.SupressExceptionWhenOrchestrationIsAlreadyRunning(async _ =>
          instanceId = await _.StartNewAsync("V1_OrchestrationTrigger", id, payload));
```

**Comment by sebastianburckhardt** [comment]
I agree that the "duplicate filtering" scenario is an important one, so we should make it easy to use. After all, we are told all the time to design our microservice APIs to provide operations that are idempotent. We could make this so easy, why don't we?

**Comment by olitomlinson** [comment]
same problem cited in this blog that the default behaviour allows orchestrations to be started again once they are in a completed state.

https://scale-tone.github.io/2020/07/02/making-durable-function-instances-unique
